sw/device/lib/dif/dif_rv_timer.h

```diff
@@ -0,0 +1,298 @@
+ * but we want a counter that seems to tick every microsecond (1MHz),
+ * we can achieve this with a prescale of 0 (so that there is a tick per
+ * clock cycle) and a tick step of 20 (since 20 * 50kHz = 1MHz).
+ *
+ * The return value of this function is only an approximation, and the
+ * actual counter frequency ultimately depends on the accuracy of the
+ * clock. The function will return an error if it cannot produce an acceptably
+ * accurate counter frequency using the given clock resolution.
+ *
+ * @param clock_freq The device clock frequency, in Hertz.
+ * @param counter_freq The desired counter frequency, in Hertz.
+ * @param out Tick parameters that will approximately produce the desired
+ *         counter frequency.
+ * @return The result of the operation.
+ */
+dif_rv_timer_approximate_tick_params_result_t
+dif_rv_timer_approximate_tick_params(uint64_t clock_freq, uint64_t counter_freq,
+                                     dif_rv_timer_tick_params_t *out);
+
+/**
+ * Starts or stops a particular hart's counter.
+ *
+ * While a counter is enabled, the counter value will increase each tick, but
+ * its timekeeping values cannot be reconfigured.
+ *
+ * @param timer A timer device.
+ * @param hart_id The hart counter to enable/disable.
+ * @param state The new enablement state.
+ * @return The result of the operation.
+ */
+dif_rv_timer_result_t dif_rv_timer_counter_set_enabled(
+    const dif_rv_timer_t *timer, uint32_t hart_id,
+    dif_rv_timer_enabled_t state);
+
+/**
+ * Reads the current value on a particlar hart's timer.
+ *
+ * @param timer A timer device.
+ * @param hart_id The hart counter to read.
+ * @param out Out param for the counter value.
+ * @return The result of the operation.
+ */
+dif_rv_timer_result_t dif_rv_timer_counter_read(const dif_rv_timer_t *timer,
+                                                uint32_t hart_id,
+                                                uint64_t *out);
+
+/**
+ * Arms the timer to go off once the counter value is greater than `threshold`,
+ * by setting up the given comparator.
+ *
+ * Beware that the following native implementation of setting an alarm
+ * contains a bug:
+ *   uint64_t time;
+ *   dif_rv_timer_counter_read(my_timer, kMyHart, &time);
+ *   time += kSomeDuration;  // (*)
+ *   dif_rv_timer_arm(my_timer, kMyHart, kMyComp, time);
+ *
+ * If `time` overflows when performing the addition, an interrupt will be fired
+ * immediately. Care should be taken to perform saturating addition at (*), so
+ * that the interrupt is fired when the wraparound happens; this way, the
+ * interrupt handler can re-arm the timer for the rest of the duration.
+ *
+ * This function makes no effort to protect the caller from setting alarms in
+ * the past that would immediately fire an interrupt. It is the caller's
+ * responsibility to read the current counter value and pick a reasonable alarm
+ * threshold.
+ *
+ * To disarm a timer, simply pass `UINT64_MAX` as the threshold.
```

Comment:
Is this accurate? Timer expires when `mtime >= mtimecmp` according to the spec.

Reply:
Oh, hmm. I'll drop the comment.